dcrec/secp256k1/loadprecomputed.go

```diff
@@ -11,58 +11,81 @@ import (
 	"encoding/binary"
 	"io/ioutil"
 	"strings"
+	"sync"
 )
 
 //go:generate go run -tags gensecp256k1 genprecomps.go
 
-// loadBytePoints decompresses and deserializes the pre-computed byte points
-// used to accelerate scalar base multiplication for the secp256k1 curve.  This
-// approach is used since it allows the compile to use significantly less ram
-// and be performed much faster than it is with hard-coding the final in-memory
-// data structure.  At the same time, it is quite fast to generate the in-memory
-// data structure at init time with this approach versus computing the table.
-func loadBytePoints() error {
-	// There will be no byte points to load when generating them.
-	bp := compressedBytePoints
-	if len(bp) == 0 {
-		return nil
-	}
+// bytePointTable describes a table used to house pre-computed values for
+// accelerating scalar base multiplication.
+type bytePointTable [32][256][2]FieldVal
 
-	// Decompress the pre-computed table used to accelerate scalar base
-	// multiplication.
-	decoder := base64.NewDecoder(base64.StdEncoding, strings.NewReader(bp))
-	r, err := zlib.NewReader(decoder)
-	if err != nil {
-		return err
-	}
-	serialized, err := ioutil.ReadAll(r)
-	if err != nil {
-		return err
-	}
+// s256BytePoints houses pre-computed values used to accelerate scalar base
+// multiplication such that they are only loaded on first use.
+var s256BytePoints = func() func() *bytePointTable {
```

Comment:
I wanted to make it impossible to access the pointer without going through the `sync.Once` loading code to ensure it's loaded.  It would have to be like this for the way you're referring to:

```Go
var bytePointData *bytePointTable
var loadBytePointsOnce sync.Once

...

func mustLoadBytePoints() {
...
	bytePointData = &data
}

func s256BytePoints() *bytePointTable {
	loadBytePointsOnce.Do(mustLoadBytePoints)
	return bytePointData
}
```` 

That means one function can accidentally access the `bytePointData` global and it may or may not work depending on if the other function that actually loads it was already called.  I actually had it that way to begin with and then hit that exact case while working on optimizing some things via some rebases because previously the only way at the data was accessing it via S256() which did the loading code.

Reply:
I should also note that if you have a suggestion of a better way to do it so the raw data pointer is not available globally, I'm definitely open to it.  I wasn't super happy with the syntax here myself, but this was the only way I could think of to ensure it was impossible to accidentally access the `nil` pointer since it sets `s256BytePoints` to a closure that closes over the loading of and access to the pointer at `init` time.